### PR TITLE
fix(ci): Stop seven D1 calls from discarding their exit status

### LINKED ISSUE #785: fix(ci): Seven D1 calls discard their exit status — three of them in teardown and destroy paths
Seven D1/Wrangler calls in CI discard their exit status. Three of them are in workflows that destroy infrastructure, where a swallowed failure means acting on data that was never read.

Surfaced by a CodeRabbit review of #784's branch, which touched these lines only to change a version string. None of them is new; the PR merely put a reviewer's eyes on them.

## The findings

**A `|| echo ""` that cannot be told from "no services enabled"** — the query, the pipe and the fallback all suppress errors, so an outage, an auth failure or a malformed response all produce the same empty string as a genuinely empty stack:

```bash
ENABLED_SERVICES=$(npx wrangler@… d1 execute "$D1_DATABASE_NAME" --remote --json \
  --command "SELECT name FROM services WHERE enabled = 1" 2>/dev/null \
  | jq -r '.[0].results | map(.name) | join(",")' 2>/dev/null || echo "")
```

- [`teardown.yml:179`](../blob/main/.github/workflows/teardown.yml#L179)
- [`destroy-all.yml:154`](../blob/main/.github/workflows/destroy-all.yml#L154)
- [`teardown-snapshot.yml:152`](../blob/main/.github/workflows/teardown-snapshot.yml#L152)

The teardown then proceeds against a list it believes is empty. Capture the query's status separately, abort when it fails, and treat zero rows as an empty list **only after a successful query**.

**A success message outside the success branch** — `|| true` forces the status to zero, then the next line states the opposite of what may have happened:

```bash
npx wrangler@… d1 execute … --command "UPDATE firewall_rules SET enabled = 0, …" 2>/dev/null || true
echo "✅ Firewall rules reset - all ports closed for next Spin Up"
```

- [`teardown.yml:296`](../blob/main/.github/workflows/teardown.yml#L296)
- [`teardown-snapshot.yml:423`](../blob/main/.github/workflows/teardown-snapshot.yml#L423)

CLAUDE.md names this exact shape under "Error Handling Principles": *"A success message outside the success branch […] an unconditional `echo "✅ done"` then states the opposite of what happened."* And the claim is a security claim — "all ports closed" — asserted without checking.

**A migration that tolerates every error, not the expected one** — [`setup-control-plane.yaml:1599`](../blob/main/.github/workflows/setup-control-plane.yaml#L1599):

```bash
npx wrangler@… d1 execute … --command "ALTER TABLE services ADD COLUMN admin_only INTEGER DEFAULT 0" 2>/dev/null || true
```

`|| true` is right for "column already exists" and wrong for a lock, a permission failure or an outage — all of which now pass silently, leaving a schema the Control Plane assumes is present. Capture stderr and tolerate only the duplicate-column message.

**A secret deletion that cannot distinguish absent from unreachable** — [`setup-control-plane.yaml:1472`](../blob/main/.github/workflows/setup-control-plane.yaml#L1472), for `BASE_DOMAIN` and `TEMPLATE_VERSION`. "Already gone" is a fine reason to continue; an API or auth error is not.

## Why this is filed rather than fixed in #784

#784 pins a version string. These are behaviour changes in teardown and destroy paths, and the only honest test is a real teardown against real infrastructure — several runs, since the interesting cases are failure cases that do not occur on demand. Bundling that into a pinning PR would make a mechanical change unreviewable and put destructive-path edits somewhere nobody would look for them.

## Note on the shape of the fix

CLAUDE.md is specific about how not to do this, in a way that is easy to reproduce by accident:

> The failure indicator is the exit status — never a by-product. […] A command that exits non-zero while writing nothing to either stream leaves `SAVE_ERROR` empty, and the failure becomes indistinguishable from success.

So: separate `*_FAILED` status from `*_ERROR` message, and print `${ERR:-(no output)}` so the silent case still says something.

## Scope check worth doing at the same time

These seven came from the files #784 happened to touch. `grep -rn '|| true\|2>/dev/null' .github/workflows/` will find more; each needs the same question — is this tolerating one expected condition, or every possible one?

### PR DESCRIPTION
**Pull request by stefanko-ch**
Closes #785

## Local CodeRabbit round

Reviewed the first commit — **3 findings, all valid, fixed in `620ac06`**. All three were the same defect in the code I had just written: the new failure branches emitted a diagnostic and then returned zero, so the workflow stayed green while the thing it reported had failed. A warning nobody's CI acts on is a quieter version of `|| true`. Detail below.

## The seven call sites

| Defect | Where | Fix |
|---|---|---|
| `\|\| echo ""` on the service-list query | teardown, destroy-all, teardown-snapshot | Query status captured separately; failure aborts; an empty result is accepted **only after a successful query** |
| `\|\| true` followed by `✅ all ports closed` | teardown, teardown-snapshot | Success message moved inside the success branch; failure path exits non-zero |
| `ALTER TABLE … \|\| true` ×3 | setup-control-plane | `migrate_column` tolerates **only** "duplicate column" |
| `pages secret delete … \|\| true` ×2 | setup-control-plane | `delete_pages_secret` distinguishes "was not there" from an auth or API error |

### The service list

```bash
ENABLED_SERVICES=$(npx wrangler … --json --command "SELECT …" 2>/dev/null \
  | jq -r '…' 2>/dev/null || echo "")
```

An outage, an auth failure and a genuinely empty table all produced the same empty string — and three workflows that destroy infrastructure then proceeded against a list they believed was empty.

stderr now goes to a file rather than into the capture: wrangler prints its banner there, and folding it into stdout would break the jq parse. That is also the evidence the banner *is* on stderr — the existing jq worked, which it could not have done otherwise.

### The firewall reset

```bash
npx wrangler … --command "UPDATE firewall_rules SET enabled = 0, …" 2>/dev/null || true
echo "✅ Firewall rules reset - all ports closed for next Spin Up"
```

A security claim, printed unconditionally after a swallowed failure. CLAUDE.md names this shape directly under "Error Handling Principles".

## What the review changed, and why it took checking first

CodeRabbit asked for `exit 1` in all three failure branches. Before adopting that I checked the step order in `teardown.yml`:

```
297  Teardown Hetzner infrastructure only
308  Confirm teardown
320  Write teardown timestamp to D1   ← the firewall reset lives here
```

The destroy has already happened, so failing here undoes nothing. What it does is stop a green check from asserting "all ports closed" when they are not — and the consequence surfaces one spin-up later, as ports nobody reopened deliberately.

Adopted, with one addition the suggestion did not have: **both messages now say the teardown itself completed and the workflow must not be re-run**, the snapshot variant adding that a re-run would take a second snapshot. A red check that provokes the wrong recovery is its own defect.

`delete_pages_secret` returns 1 for the same reason: it runs before deployment, and the caller reaches it only when the operator has cleared the value and expects the runtime to fall back. A stale secret defeats that silently.

## A weakness recorded rather than hidden

The "was not there" branch matches on wrangler's error wording. That is a heuristic: if the wording changes, a benign no-op lands in the failure branch and breaks setup.

That is the right direction to fail in — a broken setup is noticed immediately, a stale secret is not — but the comment says the fix is then to widen the pattern, not to restore `|| true`. The outcome-based alternative (delete, then list to confirm absence) would be better still and is noted as such.

## Verification

These paths cannot be triggered on demand in CI, so the shell logic was exercised against stubbed wrangler exit codes — four cases each, including the one that matters most:

```
migrate_column        ok / duplicate column / auth error / exits non-zero silently
ENABLED_SERVICES      rows / empty table / query down / unparseable response
delete_pages_secret   deleted / absent / auth error / exits non-zero silently
```

The **empty-table** and **silent-failure** cases are the point. Before this change both were indistinguishable from success, and both helpers now print `${err:-(no output)}` so a command that fails while writing nothing still says something — the failure mode CLAUDE.md describes under "The failure indicator is the exit status — never a by-product".

Behaviour under `bash -e` was checked too: the absent case continues, the auth case aborts the step with exit 1, and the line after it never runs.

## Left alone deliberately

The remaining `|| true` in these files are the cases CLAUDE.md lists as acceptable: `tofu state rm` for entries that may not exist, `grep` (which exits 1 on no match), and a `cat` for logging.

## Testing

No unit tests cover workflow shell. The real exercise is a teardown and a setup run — the firewall-reset path in particular only shows its new behaviour when D1 is unreachable, which is not something to arrange on purpose. What can be confirmed on an ordinary run is that nothing regressed: the service list still reads, the migrations still apply, and the secret cleanup still reports.

## Summary by Sourcery

Make CI workflow failures propagate instead of being discarded, while preserving explicitly tolerated no-op cases and improving diagnostics.

Bug Fixes:
- Preserve and surface failures when teardown workflows cannot read enabled services or reset firewall rules, preventing successful runs from masking unsafe state.
- Make control-plane migrations and Pages secret cleanup fail on unexpected errors while continuing only for expected already-applied or absent-resource cases.

Enhancements:
- Improve workflow diagnostics for empty command output and distinguish genuine empty results from failed D1 queries.
- Add workflow-expression tests that enforce non-empty fallbacks for captured error diagnostics and validate their detection scope.

Tests:
- Add regression coverage for captured diagnostic fallbacks, including silent failures, empty defaults, and unguarded expansions.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved deployment and teardown workflows to distinguish failed service lookups from valid empty results.
  - Added validation for malformed service data and clearer error reporting.
  - Workflows now stop safely when infrastructure updates, secret removal, or database migrations fail unexpectedly.
  - Firewall reset operations report failure accurately instead of indicating success when updates do not complete.

### CODE REVIEW

File: .github/workflows/destroy-all.yml [hunk 1]
```diff
@@ -173,14 +173,42 @@ jobs:
...
+          D1_SERVICES_JSON=""
+          if ! D1_SERVICES_JSON=$(npx wrangler@4.129.0 d1 execute "$D1_DATABASE_NAME" --remote --json \
+                --command "SELECT name FROM services WHERE enabled = 1" 2>/tmp/d1-services.err); then
+            D1_QUERY_FAILED=1
+          fi
+
+          if [ "$D1_QUERY_FAILED" -ne 0 ]; then
+            echo "❌ Could not read the enabled services from D1."
+            echo "   Continuing would act on a service list this step never obtained."
+            echo "   wrangler stderr: $(tr '\n\r' '  ' < /tmp/d1-services.err 2>/dev/null | head -c 400)"
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply .github/workflows/destroy-all.yml]: Fixed in 84de37e — all eight sites, exactly as listed.

Each printed `wrangler stderr: $(… < /tmp/*.err …)`. A command that exits non-zero while writing nothing to stderr leaves the label with nothing after it, so the log said only `wrangler stderr:` at the one moment somebody needed it. CLAUDE.md names this shape and prescribes `${VAR:-(no output)}`; the capture now goes through a variable and prints with that fallback.

Also adds three guardrails to `tests/unit/test_workflow_expressions.py`, scoped to `run:` bodies with comment lines dropped: no echo may interpolate a `.err` capture directly (that form has nowhere to put a fallback); a variable holding such a capture is never expanded bare; and the rules are not looking at an empty set.

One correction worth recording: the first version of the pattern matched `.errors[0].message` inside a jq filter and reported two lines in `setup-control-plane.yaml` that already carry a `// "unknown"` fallback. `.err` is now matched as a whole word.

All three verified under mutation, each failing on the mutated line — inline form restored, `${D1_ERR:-…}` reduced to `$D1_ERR`, and `_ERR_FILE` narrowed to match nothing.

File: .github/workflows/destroy-all.yml [hunk 2]
```diff
@@ -173,14 +173,44 @@ jobs:
...
+            D1_ERR=$(tr '\n\r' '  ' < /tmp/d1-services.err 2>/dev/null | head -c 400)
+            echo "   wrangler stderr: ${D1_ERR:-(no output)}"
+            exit 1
+          fi
+
+          if ! ENABLED_SERVICES=$(printf '%s' "$D1_SERVICES_JSON" \
+                | jq -r '.[0].results | map(.name) | join(",")'); then
+            echo "❌ D1 answered but the response could not be parsed."
+            D1_ERR=$(tr '\n\r' '  ' < /tmp/d1-services.err 2>/dev/null | head -c 400)
+            echo "   wrangler stderr: ${D1_ERR:-(no output)}"
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply .github/workflows/destroy-all.yml]: Fixed in f63491b — all three workflows.

Confirmed by reading the branch: reaching the parse block means the wrangler call already exited zero, so `/tmp/d1-services.err` holds nothing but its banner. The message pointed at the one component that had not failed, while jq's actual error went to the step log unlabelled. jq now writes to `/tmp/d1-parse.err` and the branch reports that, under a `jq stderr:` label rather than `wrangler stderr:`.

File: tests/unit/test_workflow_expressions.py [hunk 3]
```diff
@@ -358,3 +358,102 @@ def test_every_job_touching_npm_caches_the_store() -> None:
...
+        lines = _script_lines(path)
+        captured = {match.group("var") for _, line in lines if (match := _ERR_CAPTURE.match(line))}
+        for i, line in lines:
+            if "echo" not in line:
+                continue
+            for var in captured:
+                # `${VAR:-…}` is the only accepted expansion. A bare
+                # `$VAR` or a plain `${VAR}` prints nothing when empty.
+                bare = re.search(rf"\$\{{{var}\}}|\${var}\b", line)
+                guarded = f"${{{var}:-" in line
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply tests/unit/test_workflow_expressions.py]: Fixed in f63491b — the criticism is exact, and this is the sixth time in this repo a guardrail has shipped without covering its own evasion.

The check asked whether `${VAR:-` appeared anywhere on the line, so both of these passed it:

```bash
echo "   wrangler stderr: ${D1_ERR:-}"                     # prints nothing
echo "   wrangler stderr: $D1_ERR ${D1_ERR:-(no output)}"  # first half prints nothing
```

It now finds each expansion of the variable independently (new `_expansions` helper) and requires a non-empty default on every one. Both forms verified to fail the rule, each naming the offending expansion rather than just the line.

One deviation: I did not restrict it to the literal `(no output)`. The property that matters is that something is printed, and `${ERR:-none}` satisfies it; pinning the exact wording would reject a valid variant without making any failure mode less likely. Eight parametrised regression cases pin the accepted and rejected forms — including the two above, `${ERR}`, and `${ERR_COUNT}`, which is a different variable that merely starts with the same name.
